### PR TITLE
Some partial changes (don't compile yet) towards a more efficient LSTM...

### PR DESCRIPTION
**Pull request by danpovey**
implementation.
This is for @kangshiyin to look at, as I'll be asking him to help with the GPU aspects of this.
This relates to Issue #1089

### CODE REVIEW

**Comment by kangshiyin** [comment]
Thanks, I will start working on this.

File: src/cudamatrix/cu-math.cc [hunk 1]
```diff
@@ -329,14 +330,11 @@ void BackpropLstmNonlinearity(const CuMatrixBase<BaseFloat> &input,
                  self_repair_sum_out != NULL && SameDim(params, *params_deriv) &&
                  value_sum_out->NumRows() == 5 &&
                  value_sum_out->NumCols() == cell_dim &&
-                 SameDim(* ...
-                         // HERE
+                 SameDim(*value_sum_out, *deriv_sum_out) &&
+                 self_repair_sum_out->NumRows() == 5 &&
+                 value_sum_out->NumCols() == cell_dim);
```
Comment (kangshiyin) [inline src/cudamatrix/cu-math.cc]: Should it be `self_repair_sum_out->NumCols() == cell_dim` ?

**Comment by danpovey** [comment]
You're right, will fix.

**Comment by danpovey** [comment]
I just checked in some test code for the derivatives in the LSTM
backpropagation code.  The CPU version tests correctly (first time!).
However it does not test the self-repair.  I plan to check various
diagnostics in the actual training to check that the self-repair is acting
as it normally does.

If you can find some way to at least check that the CPU and GPU versions
are doing the same thing w.r.t. the self-repair and stats accumulation
code, it would be nice.

**Comment by danpovey** [comment]
I'm concerned about the effects of roundoff if it's converted to
BaseFloat.  And other value-sum/deriv-sum quantities are done in double.
So it's best if you leave it at double, but the quantities that you compute
internally can be in BaseFloat (and only convert to double when you add to
the deriv_sum and value_sum).

**Comment by kangshiyin** [comment]
ok, I will follow your type definitions.

**Comment by danpovey** [comment]
Note, this will not be merged, @kangshiyin's PR #1206 is the version that will be merged.  But leaving this up for now as we might learn something from the automated checks.

**Comment by danpovey** [comment]
This PR will be merged indirectly via PR #1229.

**Pull request closed without merging**